gap/bliss.gi

```diff
@@ -12,141 +13,247 @@
 
 InstallMethod(DigraphCanonicalLabelling, "for a digraph",
 [IsDigraph],
-function(graph)
-
-  if IsMultiDigraph(graph) then
-    return MULTIDIGRAPH_CANONICAL_LABELING(graph);
+function(digraph)
+  if IsMultiDigraph(digraph) then
+    return MULTIDIGRAPH_CANONICAL_LABELING(digraph);
   fi;
-  return DIGRAPH_CANONICAL_LABELING(graph);
+  return DIGRAPH_CANONICAL_LABELING(digraph);
 end);
 
-InstallMethod(DigraphCanonicalLabelling,
-"for a digraph and homogeneous list",
+#
+
+InstallMethod(DigraphCanonicalLabelling, "for a digraph and homogeneous list",
 [IsDigraph, IsHomogeneousList],
-function(graph, list)
-  local colors, i;
+function(digraph, list)
+  local n, colours;
 
-  if IsMultiDigraph(graph) then
+  if IsMultiDigraph(digraph) then
     return fail;
   fi;
-  if (not IsEmpty(list)) and IsList(list[1]) then # color classes
-    colors := [1 .. DigraphNrVertices(graph)];
-    if not (ForAll(list, IsDuplicateFreeList) and Union(list) = colors) then
-      ErrorNoReturn("Digraphs: DigraphCanonicalLabelling: usage,\n",
-                    "the union of the lists in the second arg should equal ",
-                    "[1 .. ", DigraphNrVertices(graph), "],");
-    fi;
-
-    for i in [1 .. Length(list)] do
-      colors{list[i]} := [1 .. Length(list[i])] * 0 + i;
-    od;
-  else
-    if not (Length(list) = DigraphNrVertices(graph)
-            and ForAll(list, c -> IsPosInt(c) and 1 <= c
-                                  and c <= DigraphNrVertices(graph))) then
-      ErrorNoReturn("Digraphs: DigraphCanonicalLabelling: usage,\n",
-                    "the second arg must be a list of length ",
-                    DigraphNrVertices(graph), " of integers in [1 .. ",
-                    DigraphNrVertices(graph), "],");
-    fi;
-    colors := list;
+  n := DigraphNrVertices(digraph);
+  colours := DIGRAPHS_ValidateVertexPartition(n, list);
+  if colours = fail then
+    ErrorNoReturn("Digraphs: DigraphCanonicalLabelling: usage,\n",
+                  "the argument <list> does not define a valid colouring of ",
+                  "the vertices of\n<digraph>. Either <list> should be a list ",
+                  "of length <n> consisting of integers\nin the range ",
+                  "[0 .. <n>], or <list> should be a list of non-empty ",
+                  "disjoint\nlists whose union is the vertices of <digraph>,");
```

Comment:
I think this error message could be more helpful. What is `n` for example? I think it would be better to add this error message to `DIGRAPHS_ValidateVertexPartition`and be more specific about what is wrong (as was previously attempted). I haven't yet looked at `DIGRAPHS_ValidateVertexPartition`, will comment more there.